File: disc_golf_app/lib/models/roulette_data.dart
import 'dart:math';

/// Putting styles for when a player is in putting range.
enum PuttStyle {
  kneeling,
  straddle,
  turbo,
  spin,
  push,
  spush,
}

enum ShotType {
  hyzer,
  anhyzer,
  flat,
  roller,
  tomahawk,
  thumber,
  grenade,
  scoober,
}

enum PowerModifier {
  fullPower,
  halfPower,
  quarterPower,
  overhand,
  standstill,
  runUp,
  xStep,
}

enum Hindrance {
  none,
  offHand,
  eyesClosed,
  backwards,
  oneLeg,
  sitting,
  kneeling,
  spinFirst,
}

/// Disc lists for different game modes.
class DiscLists {
  /// Full disc list for casual roulette.
  static const all = [
    'Putter', 'Approach', 'Utility', 'Midrange', 'Fairway Driver', 'Distance Driver',
  ];

  /// Speed 4 and under — used in scored rounds to encourage actual scoring.
  static const scoringRound = ['Putter', 'Approach', 'Midrange'];

  /// Putters only — used when in putting range.
  static const putting = ['Putter'];
}

class RouletteResult {
  final ShotType shotType;
  final String? discName;
  final PowerModifier powerModifier;
  final Hindrance hindrance;
  final PuttStyle? puttStyle; // Non-null when this is a putting challenge
  final DateTime timestamp;
  
  double getDifficultyMultiplier() {
  double multiplier = 1.0;
  
  // Shot type difficulty
  switch (shotType) {
    case ShotType.flat:
    case ShotType.hyzer:
      multiplier += 0.0;
      break;
    case ShotType.anhyzer:
      multiplier += 0.2;
      break;
    case ShotType.roller:
      multiplier += 0.3;
      break;
    case ShotType.scoober:
    case ShotType.tomahawk:
    case ShotType.thumber:
      multiplier += 0.5;
      break;
    case ShotType.grenade:
      multiplier += 0.7;
      break;
  }
  
  // Power modifier difficulty
  switch (powerModifier) {
    case PowerModifier.fullPower:
    case PowerModifier.runUp:
    case PowerModifier.xStep:
      multiplier += 0.0;
      break;
    case PowerModifier.halfPower:
      multiplier += 0.1;
      break;
    case PowerModifier.standstill:
      multiplier += 0.2;
      break;
    case PowerModifier.quarterPower:
      multiplier += 0.3;
      break;
    case PowerModifier.overhand:
      multiplier += 0.4;
      break;
  }
  
  // Hindrance difficulty (biggest impact)
  switch (hindrance) {
    case Hindrance.none:
      multiplier += 0.0;
      break;
    case Hindrance.kneeling:
    case Hindrance.sitting:
      multiplier += 0.3;
      break;
    case Hindrance.oneLeg:
      multiplier += 0.5;
      break;
    case Hindrance.spinFirst:
      multiplier += 0.7;
      break;
    case Hindrance.offHand:
      multiplier += 1.0;
      break;
    case Hindrance.backwards:
      multiplier += 1.2;
      break;
    case Hindrance.eyesClosed:
      multiplier += 1.5;
      break;
  }
  
  // Putt style difficulty (when putting)
  if (puttStyle != null) {
    switch (puttStyle!) {
      case PuttStyle.push:
        multiplier += 0.0;
        break;
      case PuttStyle.spush:
        multiplier += 0.1;
        break;
      case PuttStyle.spin:
        multiplier += 0.2;
        break;
      case PuttStyle.straddle:
        multiplier += 0.3;
        break;
      case PuttStyle.kneeling:
        multiplier += 0.5;
        break;
      case PuttStyle.turbo:
        multiplier += 0.7;
        break;
    }
  }

  return multiplier;
}

factory RouletteResult.fromJson(Map<String, dynamic> json) {
  return RouletteResult(
    shotType: ShotType.values.firstWhere((e) => e.name == json['shotType']),
    discName: json['discName'],
    powerModifier: PowerModifier.values.firstWhere((e) => e.name == json['powerModifier']),
    hindrance: Hindrance.values.firstWhere((e) => e.name == json['hindrance']),
    puttStyle: json['puttStyle'] != null
        ? PuttStyle.values.firstWhere((e) => e.name == json['puttStyle'])
        : null,
    timestamp: DateTime.parse(json['timestamp']),
  );
}

  RouletteResult({
    required this.shotType,
    this.discName,
    required this.powerModifier,
    required this.hindrance,
    this.puttStyle,
    required this.timestamp,
  });

  static RouletteResult generate(List<String> availableDiscs) {
    final random = Random();

    return RouletteResult(
      shotType: ShotType.values[random.nextInt(ShotType.values.length)],
      discName: availableDiscs.isNotEmpty
          ? availableDiscs[random.nextInt(availableDiscs.length)]
          : null,
      powerModifier: PowerModifier.values[random.nextInt(PowerModifier.values.length)],
      hindrance: Hindrance.values[random.nextInt(Hindrance.values.length)],
      timestamp: DateTime.now(),
    );
  }

  /// Generate a putting-specific challenge.
  static RouletteResult generatePutt(List<String> availableDiscs) {
    final random = Random();

    return RouletteResult(
      shotType: ShotType.flat, // Putts are always flat
      discName: availableDiscs.isNotEmpty
          ? availableDiscs[random.nextInt(availableDiscs.length)]
          : null,
      powerModifier: PowerModifier.standstill, // Putts are standstill
      hindrance: Hindrance.none, // Hindrance replaced by putt style
      puttStyle: PuttStyle.values[random.nextInt(PuttStyle.values.length)],
      timestamp: DateTime.now(),
    );
  }

  bool get isPutt => puttStyle != null;

  String getPuttStyleDescription() {
    if (puttStyle == null) return '';
    switch (puttStyle!) {
      case PuttStyle.push:
        return 'Push Putt - Wrist stays firm';
      case PuttStyle.spush:
        return 'Spush Putt - Spin + push hybrid';
      case PuttStyle.spin:
        return 'Spin Putt - Wrist flick';
      case PuttStyle.straddle:
        return 'Straddle Putt - Feet parallel to basket';
      case PuttStyle.kneeling:
        return 'Kneeling Putt - One knee down';
      case PuttStyle.turbo:
        return 'Turbo Putt - Overhead grip';
    }
  }

  String getShotTypeDescription() {
    switch (shotType) {
      case ShotType.hyzer:
        return 'Hyzer - Outside edge down';
      case ShotType.anhyzer:
        return 'Anhyzer - Outside edge up';
      case ShotType.flat:
        return 'Flat - Level release';
      case ShotType.roller:
        return 'Roller - Ground roll shot';
      case ShotType.tomahawk:
        return 'Tomahawk - Overhead hyzer';
      case ShotType.thumber:
        return 'Thumber - Overhead anhyzer';
      case ShotType.grenade:
        return 'Grenade - Vertical release';
      case ShotType.scoober:
        return 'Scoober - Upside down';
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'shotType': shotType.name,
      'discName': discName,
      'powerModifier': powerModifier.name,
      'hindrance': hindrance.name,
      'puttStyle': puttStyle?.name,
      'timestamp': timestamp.toIso8601String(),
    };
  }
}

class GameSession {
  final String id;
  final List<String> players;
  final List<RouletteResult> results;
  final DateTime startedAt;
  DateTime? endedAt;

  GameSession({
    required this.id,
    required this.players,
    required this.results,
    required this.startedAt,
    this.endedAt,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'players': players,
      'results': results.map((r) => r.toJson()).toList(),
      'startedAt': startedAt.toIso8601String(),
      'endedAt': endedAt?.toIso8601String(),
    };
  }
}

File: disc_golf_android/core/src/main/kotlin/com/discflightschool/core/model/RouletteData.kt
package com.discflightschool.core.model

import com.discflightschool.core.util.DateTimes
import java.time.Instant
import kotlin.random.Random
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** Putting styles for when a player is in putting range. */
enum class PuttStyle {
    KNEELING, STRADDLE, TURBO, SPIN, PUSH, SPUSH;

    /** The name as persisted by earlier versions of the app (Dart enum names). */
    val jsonName: String
        get() = when (this) {
            KNEELING -> "kneeling"
            STRADDLE -> "straddle"
            TURBO -> "turbo"
            SPIN -> "spin"
            PUSH -> "push"
            SPUSH -> "spush"
        }

    companion object {
        fun fromJson(name: String): PuttStyle = entries.first { it.jsonName == name }
    }
}

enum class ShotType {
    HYZER, ANHYZER, FLAT, ROLLER, TOMAHAWK, THUMBER, GRENADE, SCOOBER;

    val jsonName: String
        get() = when (this) {
            HYZER -> "hyzer"
            ANHYZER -> "anhyzer"
            FLAT -> "flat"
            ROLLER -> "roller"
            TOMAHAWK -> "tomahawk"
            THUMBER -> "thumber"
            GRENADE -> "grenade"
            SCOOBER -> "scoober"
        }

    companion object {
        fun fromJson(name: String): ShotType = entries.first { it.jsonName == name }
    }
}

enum class PowerModifier {
    FULL_POWER, HALF_POWER, QUARTER_POWER, OVERHAND, STANDSTILL, RUN_UP, X_STEP;

    val jsonName: String
        get() = when (this) {
            FULL_POWER -> "fullPower"
            HALF_POWER -> "halfPower"
            QUARTER_POWER -> "quarterPower"
            OVERHAND -> "overhand"
            STANDSTILL -> "standstill"
            RUN_UP -> "runUp"
            X_STEP -> "xStep"
        }

    val displayName: String
        get() = when (this) {
            FULL_POWER -> "Full Power"
            HALF_POWER -> "Half Power"
            QUARTER_POWER -> "Quarter Power"
            OVERHAND -> "Overhand"
            STANDSTILL -> "Standstill"
            RUN_UP -> "Run Up"
            X_STEP -> "X Step"
        }

    companion object {
        fun fromJson(name: String): PowerModifier = entries.first { it.jsonName == name }
    }
}

enum class Hindrance {
    NONE, OFF_HAND, EYES_CLOSED, BACKWARDS, ONE_LEG, SITTING, KNEELING, SPIN_FIRST;

    val jsonName: String
        get() = when (this) {
            NONE -> "none"
            OFF_HAND -> "offHand"
            EYES_CLOSED -> "eyesClosed"
            BACKWARDS -> "backwards"
            ONE_LEG -> "oneLeg"
            SITTING -> "sitting"
            KNEELING -> "kneeling"
            SPIN_FIRST -> "spinFirst"
        }

    val displayName: String
        get() = when (this) {
            NONE -> "No Hindrance"
            OFF_HAND -> "Off Hand"
            EYES_CLOSED -> "Eyes Closed"
            BACKWARDS -> "Backwards"
            ONE_LEG -> "One Leg"
            SITTING -> "Sitting"
            KNEELING -> "Kneeling"
            SPIN_FIRST -> "Spin First"
        }

    companion object {
        fun fromJson(name: String): Hindrance = entries.first { it.jsonName == name }
    }
}

/** Disc lists for the different game modes. */
object DiscLists {
    /** The full disc list for casual roulette. */
    val all = listOf(
        "Putter", "Approach", "Utility", "Midrange", "Fairway Driver", "Distance Driver",
    )

    /** Speed 4 and under — used in scored rounds to encourage actual scoring. */
    val scoringRound = listOf("Putter", "Approach", "Midrange")

    /** Putters only — used when in putting range. */
    val putting = listOf("Putter")
}

/** One spun challenge: what to throw, how to throw it, and what to do to yourself first. */
data class RouletteResult(
    val shotType: ShotType,
    val discName: String? = null,
    val powerModifier: PowerModifier,
    val hindrance: Hindrance,
    /** Non-null when this is a putting challenge. */
    val puttStyle: PuttStyle? = null,
    val timestamp: Instant,
) {
    val isPutt: Boolean get() = puttStyle != null

    val difficultyMultiplier: Double
        get() {
            var multiplier = 1.0

            multiplier += when (shotType) {
                ShotType.FLAT, ShotType.HYZER -> 0.0
                ShotType.ANHYZER -> 0.2
                ShotType.ROLLER -> 0.3
                ShotType.SCOOBER, ShotType.TOMAHAWK, ShotType.THUMBER -> 0.5
                ShotType.GRENADE -> 0.7
            }

            multiplier += when (powerModifier) {
                PowerModifier.FULL_POWER, PowerModifier.RUN_UP, PowerModifier.X_STEP -> 0.0
                PowerModifier.HALF_POWER -> 0.1
                PowerModifier.STANDSTILL -> 0.2
                PowerModifier.QUARTER_POWER -> 0.3
                PowerModifier.OVERHAND -> 0.4
            }

            // Hindrance has the biggest impact.
            multiplier += when (hindrance) {
                Hindrance.NONE -> 0.0
                Hindrance.KNEELING, Hindrance.SITTING -> 0.3
                Hindrance.ONE_LEG -> 0.5
                Hindrance.SPIN_FIRST -> 0.7
                Hindrance.OFF_HAND -> 1.0
                Hindrance.BACKWARDS -> 1.2
                Hindrance.EYES_CLOSED -> 1.5
            }

            puttStyle?.let {
                multiplier += when (it) {
                    PuttStyle.PUSH -> 0.0
                    PuttStyle.SPUSH -> 0.1
                    PuttStyle.SPIN -> 0.2
                    PuttStyle.STRADDLE -> 0.3
                    PuttStyle.KNEELING -> 0.5
                    PuttStyle.TURBO -> 0.7
                }
            }

            return multiplier
        }

    val shotTypeDescription: String
        get() = when (shotType) {
            ShotType.HYZER -> "Hyzer - Outside edge down"
            ShotType.ANHYZER -> "Anhyzer - Outside edge up"
            ShotType.FLAT -> "Flat - Level release"
            ShotType.ROLLER -> "Roller - Ground roll shot"
            ShotType.TOMAHAWK -> "Tomahawk - Overhead hyzer"
            ShotType.THUMBER -> "Thumber - Overhead anhyzer"
            ShotType.GRENADE -> "Grenade - Vertical release"
            ShotType.SCOOBER -> "Scoober - Upside down"
        }

    val puttStyleDescription: String
        get() = when (puttStyle) {
            null -> ""
            PuttStyle.PUSH -> "Push Putt - Wrist stays firm"
            PuttStyle.SPUSH -> "Spush Putt - Spin + push hybrid"
            PuttStyle.SPIN -> "Spin Putt - Wrist flick"
            PuttStyle.STRADDLE -> "Straddle Putt - Feet parallel to basket"
            PuttStyle.KNEELING -> "Kneeling Putt - One knee down"
            PuttStyle.TURBO -> "Turbo Putt - Overhead grip"
        }

    fun toJson(): JsonObject = buildJsonObject {
        put("shotType", JsonPrimitive(shotType.jsonName))
        put("discName", discName?.let { JsonPrimitive(it) } ?: JsonNull)
        put("powerModifier", JsonPrimitive(powerModifier.jsonName))
        put("hindrance", JsonPrimitive(hindrance.jsonName))
        put("puttStyle", puttStyle?.let { JsonPrimitive(it.jsonName) } ?: JsonNull)
        put("timestamp", JsonPrimitive(DateTimes.format(timestamp)))
    }

    companion object {
        /**
         * Power modifiers that are physically impossible with a given hindrance.
         * Hindrance is picked first; incompatible powers are removed before sampling.
         */
        private val incompatiblePower: Map<Hindrance, List<PowerModifier>> = mapOf(
            // You cannot run or x-step from the ground.
            Hindrance.KNEELING to listOf(PowerModifier.RUN_UP, PowerModifier.X_STEP),
            Hindrance.SITTING to listOf(PowerModifier.RUN_UP, PowerModifier.X_STEP),
            // You cannot plant both feet for an x-step, or run, on one leg.
            Hindrance.ONE_LEG to listOf(PowerModifier.RUN_UP, PowerModifier.X_STEP),
        )

        fun generate(
            availableDiscs: List<String>,
            random: Random = Random.Default,
            now: Instant = Instant.now(),
        ): RouletteResult {
            val hindrance = Hindrance.entries[random.nextInt(Hindrance.entries.size)]
            val blocked = incompatiblePower[hindrance].orEmpty()
            val validPowers = PowerModifier.entries.filterNot { it in blocked }

            return RouletteResult(
                shotType = ShotType.entries[random.nextInt(ShotType.entries.size)],
                discName = availableDiscs.getOrNull(
                    if (availableDiscs.isEmpty()) -1 else random.nextInt(availableDiscs.size),
                ),
                powerModifier = validPowers[random.nextInt(validPowers.size)],
                hindrance = hindrance,
                timestamp = now,
            )
        }

        /** Generate a putting-specific challenge. */
        fun generatePutt(
            availableDiscs: List<String>,
            random: Random = Random.Default,
            now: Instant = Instant.now(),
        ): RouletteResult = RouletteResult(
            // Putts are always flat, always standstill, and the hindrance is
            // replaced by the putt style.
            shotType = ShotType.FLAT,
            discName = availableDiscs.getOrNull(
                if (availableDiscs.isEmpty()) -1 else random.nextInt(availableDiscs.size),
            ),
            powerModifier = PowerModifier.STANDSTILL,
            hindrance = Hindrance.NONE,
            puttStyle = PuttStyle.entries[random.nextInt(PuttStyle.entries.size)],
            timestamp = now,
        )

        fun fromJson(json: JsonObject): RouletteResult = RouletteResult(
            shotType = ShotType.fromJson(json.getValue("shotType").jsonPrimitive.content),
            discName = json["discName"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
            powerModifier = PowerModifier.fromJson(
                json.getValue("powerModifier").jsonPrimitive.content,
            ),
            hindrance = Hindrance.fromJson(json.getValue("hindrance").jsonPrimitive.content),
            puttStyle = json["puttStyle"]?.takeIf { it !is JsonNull }
                ?.jsonPrimitive?.content?.let { PuttStyle.fromJson(it) },
            timestamp = DateTimes.parse(json.getValue("timestamp").jsonPrimitive.content),
        )
    }
}

/** A casual (unscored) session of spins shared by a group. */
data class GameSession(
    val id: String,
    val players: List<String>,
    val results: List<RouletteResult>,
    val startedAt: Instant,
    val endedAt: Instant? = null,
) {
    fun toJson(): JsonObject = buildJsonObject {
        put("id", JsonPrimitive(id))
        put("players", JsonArray(players.map { JsonPrimitive(it) }))
        put("results", JsonArray(results.map { it.toJson() }))
        put("startedAt", JsonPrimitive(DateTimes.format(startedAt)))
        put("endedAt", endedAt?.let { JsonPrimitive(DateTimes.format(it)) } ?: JsonNull)
    }

    companion object {
        fun fromJson(json: JsonObject): GameSession = GameSession(
            id = json.getValue("id").jsonPrimitive.content,
            players = json.getValue("players").jsonArray.map { it.jsonPrimitive.content },
            results = json.getValue("results").jsonArray.map {
                RouletteResult.fromJson(it.jsonObject)
            },
            startedAt = DateTimes.parse(json.getValue("startedAt").jsonPrimitive.content),
            endedAt = json["endedAt"]?.takeIf { it !is JsonNull }
                ?.jsonPrimitive?.content?.let { DateTimes.parse(it) },
        )
    }
}
